File: test/marketplace_listing_lifecycle_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:pipe_app/marketplace/marketplace_listing_lifecycle.dart';

void main() {
  final now = DateTime.utc(2026, 8, 15, 12);

  test('active listing shows days remaining', () {
    final lifecycle = MarketplaceListingLifecycle.fromMap({
      'status': 'active',
      'expiresAt': now.add(const Duration(days: 12)),
    }, now: now);
    expect(lifecycle.expired, isFalse);
    expect(lifecycle.daysRemaining, 12);
    expect(lifecycle.expiringSoon, isFalse);
  });

  test('three-day window becomes expiring soon', () {
    final lifecycle = MarketplaceListingLifecycle.fromMap({
      'status': 'active',
      'expiresAt': now.add(const Duration(days: 2, hours: 2)),
    }, now: now);
    expect(lifecycle.expired, isFalse);
    expect(lifecycle.expiringSoon, isTrue);
    expect(lifecycle.daysRemaining, 3);
  });

  test('expired status remains explicit', () {
    final lifecycle = MarketplaceListingLifecycle.fromMap({
      'status': 'expired',
      'renewalCount': 2,
    }, now: now);
    expect(lifecycle.expired, isTrue);
    expect(lifecycle.renewalCount, 2);
    expect(lifecycle.ownerLabel, contains('Expired'));
  });
}

File: firebase/functions/test/marketplace_listing_lifecycle.test.js
"use strict";

const test = require("node:test");
const assert = require("node:assert/strict");
const {
  DAY_MS,
  MARKETPLACE_LISTING_ACTIVE_DAYS,
  listingExpiryMillis,
  listingLifecycleState,
} = require("../marketplace_listing_lifecycle");

const now = Date.UTC(2026, 7, 15, 12, 0, 0);

test("Marketplace listings expire after 30 days", () => {
  assert.equal(MARKETPLACE_LISTING_ACTIVE_DAYS, 30);
  assert.equal(
      listingExpiryMillis(now),
      now + 30 * DAY_MS,
  );
});

test("active listing reports time remaining", () => {
  const state = listingLifecycleState({
    status: "active",
    expiresAt: now + 12 * DAY_MS,
  }, now);
  assert.equal(state.expired, false);
  assert.equal(state.expiringSoon, false);
  assert.equal(state.remainingMs, 12 * DAY_MS);
});

test("listing enters expiring-soon window three days before expiry", () => {
  const state = listingLifecycleState({
    status: "active",
    expiresAt: now + 2 * DAY_MS,
  }, now);
  assert.equal(state.expired, false);
  assert.equal(state.expiringSoon, true);
});

test("expired status remains explicit even without an expiry timestamp", () => {
  const state = listingLifecycleState({status: "expired"}, now);
  assert.equal(state.expired, true);
});

test("elapsed active or paused listing is treated as expired", () => {
  for (const status of ["active", "paused"]) {
    const state = listingLifecycleState({
      status,
      expiresAt: now - 1,
    }, now);
    assert.equal(state.expired, true);
  }
});
